File: +fs/+lpew/+v2/preLPEW2.m
function [env, weight, s] = preLPEW2(env, parms)
%FS.LPEW.V2.PRELPEW2  Vectorized LPEW2 driver (drop-in replacement for ferncodes_Pre_LPEW_2_vect).
%
%   [env, weight, s] = fs.lpew.v2.preLPEW2(env, parms)
%
%   Runs the full LPEW2 pipeline as batched ops:
%     1. Build FS + CSR + shifts
%     2. Batched OPT → P, T, O, Qcorner
%     3. Batched ksInterp → Kt1/Kt2/Kn1/Kn2 (all corners)
%     4. Batched angulos + netas → theta1/theta2, ve1/ve2, netas (all corners)
%     5. Batched lambdaWeights → lambda (interior vect + boundary loop), r
%     6. Segmented normalisation:  weight = lambda / sum(lambda per node)
%     7. Neumann source s via env.benchmark.calcularTermoNeumannVet
%
%   Same output layout as ferncodes_Pre_LPEW_2_vect: weight is a 1×nCorners
%   row-vector aligned with esurn1 flat order. s is [nNodes x 1].

    % ── Ensure the kmap the caller expected ────────────────────────
    if env.config.numcase > 400 && isfield(parms, 'auxperm') && ~isempty(parms.auxperm)
        env.config.kmap = parms.auxperm;
    elseif ~isfield(env.config, 'kmap') || isempty(env.config.kmap)
        env.config.kmap = env.config.perm;
    end

    % ── Build FS scaffolding ───────────────────────────────────────
    FS = fs.mesh.build(env);
    FS = fs.csr.buildCorners(FS);
    FS = fs.csr.buildCornerShifts(FS);
    FS.cfg.phasekey = env.config.phasekey;

    % ── Batched kernels ────────────────────────────────────────────
    [P_all, T_all, O_all, Qc_all, ~] = fs.lpew.OPT(FS); %#ok<ASGLU>
    [Kt1, Kt2, Kn1, Kn2] = fs.lpew.v2.ksInterp(FS, T_all, Qc_all);
    [ve2, ve1, theta2, theta1] = fs.lpew.v2.angulos(FS, T_all, O_all, Qc_all);
    netas_all = fs.lpew.v2.netas(FS, T_all, P_all, O_all, Qc_all);
    [lambda, r] = fs.lpew.v2.lambdaWeights(FS, Kt1, Kt2, Kn1, Kn2, ...
        theta1, theta2, ve1, ve2, netas_all, T_all, Qc_all);

    % ── Segmented normalisation: weight = lambda / sum(lambda per node) ─
    sum_lambda = accumarray(FS.csr.cornerNode, lambda, [FS.mesh.nNodes, 1]);
    % Guard against 0/0 (isolated nodes with no lambda contribution — shouldn't
    % happen on a well-formed mesh but be defensive)
    denom = sum_lambda(FS.csr.cornerNode);
    denom(denom == 0) = 1;
    weight_col = lambda ./ denom;

    % Row-vector layout matching legacy Pre_LPEW_2_vect output
    weight = weight_col.';

    % ── Neumann source term (benchmark-provided) ───────────────────
    N = env.premethod.MPFAD.N;
    s = env.benchmark.calcularTermoNeumannVet(r, sum_lambda, N, env);

    % ── Persist in env.premethod ───────────────────────────────────
    env.premethod.MPFAD.weight = weight;
    env.premethod.MPFAD.s      = s;
end

File: fs/lpew/v2/preLPEW2.py
# -*- coding: utf-8 -*-
"""
fs.lpew.v2.preLPEW2 (Python) — tradução de +fs/+lpew/+v2/preLPEW2.m

Pipeline LPEW2 vetorizado:
- constrói estruturas FS + CSR + shifts
- executa kernels batched: OPT, ksInterp, angulos, netas, lambdaWeights
- normaliza lambda por nó para obter weight (1 x nCorners)
- calcula termo de Neumann s via benchmark.calcularTermoNeumannVet
- persiste env.premethod.MPFAD.weight e env.premethod.MPFAD.s

Assunções:
- env e parms são dicts com estruturas numpy-friendly
- fs.mesh.build normaliza os índices CSR legados para 0-based

Retorna: (env, weight, s)
- weight: 1D numpy array (row) com tamanho nCorners (legacy: 1 x nCorners row-vector)
- s: 1D numpy array shape (nNodes,)
"""
from typing import Tuple, Dict, Any
import numpy as np


def preLPEW2(env: Dict[str, Any], parms: Dict[str, Any]) -> Tuple[Dict[str, Any], np.ndarray, np.ndarray]:
    # Ensure kmap as MATLAB code does
    config = env.setdefault("config", {})
    auxperm = parms.get("auxperm")
    if config.get("numcase", 0) > 400 and not _is_empty(auxperm):
        config["kmap"] = auxperm
    elif _is_empty(config.get("kmap")):
        config["kmap"] = _default_kmap(config.get("perm"))

    # Build FS scaffolding
    from fs.mesh import build as mesh_build
    from fs.csr import buildCorners, buildCornerShifts

    FS = mesh_build(env)
    FS = buildCorners(FS)
    FS = buildCornerShifts(FS)
    # Keep compatibility with MATLAB: FS.cfg.phasekey
    FS.setdefault("cfg", {})["phasekey"] = env.get("config", {}).get("phasekey")

    # Batched kernels
    from fs.lpew import OPT
    from fs.lpew.v2 import ksInterp, angulos, netas, lambdaWeights

    P_all, T_all, O_all, Qc_all, _ = OPT(FS)
    Kt1, Kt2, Kn1, Kn2 = ksInterp(FS, T_all, Qc_all)
    ve2, ve1, theta2, theta1 = angulos(FS, T_all, O_all, Qc_all)
    netas_all = netas(FS, T_all, P_all, O_all, Qc_all)
    lambda_arr, r = lambdaWeights(FS, Kt1, Kt2, Kn1, Kn2,
                                   theta1, theta2, ve1, ve2, netas_all, T_all, Qc_all)

    # Segmented normalization: weight = lambda / sum(lambda per node)
    # FS.csr.cornerNode: array length nCorners mapping each corner -> node index (0-based)
    csr = FS.get("csr")
    if csr is None or "cornerNode" not in csr:
        raise RuntimeError("FS.csr.cornerNode not found — ensure fs.csr.buildCorners produced it")

    cornerNode = np.asarray(csr["cornerNode"]).astype(int)
    nNodes = int(FS["mesh"]["nNodes"])

    # sum_lambda per node (equivalent a accumarray in MATLAB)
    sum_lambda = np.bincount(cornerNode, weights=lambda_arr, minlength=nNodes)

    # denom per corner: sum_lambda mapped by cornerNode
    denom = sum_lambda[cornerNode]
    # Guard against division by zero
    denom_safe = np.where(denom == 0, 1.0, denom)

    weight_col = np.asarray(lambda_arr) / denom_safe

    # Legacy layout: row-vector (1 x nCorners). We'll return a 1D numpy array but
    # document it as row layout (transpose if needed by callers)
    weight = weight_col.reshape(-1)

    # Neumann source term (benchmark-provided) — N is env.premethod.MPFAD.N
    N = env["premethod"]["MPFAD"]["N"]
    benchmark = env["benchmark"]
    if isinstance(benchmark, dict):
        neumann_callback = benchmark.get("calcularTermoNeumannVet")
    else:
        neumann_callback = getattr(benchmark, "calcularTermoNeumannVet", None)
    if not callable(neumann_callback):
        raise TypeError("env.benchmark.calcularTermoNeumannVet must be callable")
    s = np.asarray(neumann_callback(r, sum_lambda, N, env)).reshape(-1)

    # Persist in env.premethod
    env.setdefault("premethod", {}).setdefault("MPFAD", {})["weight"] = weight
    env.setdefault("premethod", {}).setdefault("MPFAD", {})["s"] = s

    return env, weight, s


def _is_empty(value: Any) -> bool:
    return value is None or np.asarray(value).size == 0


def _default_kmap(perm: Any) -> np.ndarray:
    if _is_empty(perm):
        raise ValueError("env.config.perm is required when kmap is empty")
    tensor = np.asarray(perm)
    if tensor.ndim == 1:
        tensor = tensor.reshape(1, -1)
    if tensor.ndim != 2:
        raise ValueError("env.config.perm must be a one- or two-dimensional array")
    if tensor.shape[1] >= 5:
        return tensor
    if tensor.shape[1] != 4:
        raise ValueError("env.config.perm must contain K11/K12/K21/K22")
    material_ids = np.arange(1, tensor.shape[0] + 1).reshape(-1, 1)
    return np.column_stack((material_ids, tensor))
